Framework/Core/include/Framework/AnalysisTask.h

```diff
@@ -392,146 +393,236 @@ struct AnalysisDataProcessorBuilder {
-
-              // for each grouping element we need to slice the selection vector
-              auto start_iterator = std::lower_bound(current_start, fullSelection.end(), offsets[oi]);
-              auto stop_iterator = std::lower_bound(start_iterator, fullSelection.end(), offsets[oi + 1]);
-              current_start = stop_iterator;
-              soa::SelectionVector slicedSelection{start_iterator, stop_iterator};
-              std::transform(slicedSelection.begin(), slicedSelection.end(), slicedSelection.begin(), [&](int64_t index) { return index - static_cast<int64_t>(offsets[oi]); });
-
-              std::decay_t<AssociatedType> typedTable{{groupedElementsTable}, std::move(slicedSelection), offsets[oi]};
-              typedTable.bindExternalIndices(&groupingTable);
-              task.process(groupingElement, typedTable);
-              ++const_cast<std::decay_t<Grouping>&>(groupingElement);
-              ++oi;
-            }
-          } else if constexpr (soa::is_soa_join_t<std::decay_t<AssociatedType>>::value || soa::is_soa_concat_t<std::decay_t<AssociatedType>>::value) {
-            for (auto& groupedDatum : groupsCollection) {
-              auto groupedElementsTable = arrow::util::get<std::shared_ptr<arrow::Table>>(groupedDatum.value);
-              // Set the refererred table.
-              // FIXME: we should be able to do this upfront for all the tables.
-              std::decay_t<AssociatedType> typedTable{{groupedElementsTable}, offsets[oi]};
-              typedTable.bindExternalIndices(&groupingTable);
-              task.process(groupingElement, typedTable);
-              ++const_cast<std::decay_t<Grouping>&>(groupingElement);
-              ++oi;
-            }
-          }
-        } else {
-          static_assert(always_static_assert_v<AssociatedType>, "I do not know how to iterate on this");
+    } else {
+      // multiple arguments to process
+      static_assert(((soa::is_soa_iterator_t<std::decay_t<Associated>>::value == false) && ...),
+                    "Associated arguments of process() should not be iterators");
+      auto associatedTables = AnalysisDataProcessorBuilder::bindAssociatedTables(inputs, &C::process, infos);
+      if constexpr (soa::is_soa_iterator_t<G>::value) {
+        // grouping case
+        auto slicer = GroupSlicer(groupingTable, associatedTables);
+        for (auto& slice : slicer) {
+          auto associatedSlices = slice.AssociatedTables();
+          (std::get<std::decay_t<Associated>>(associatedSlices).bindExternalIndices(&groupingTable), ...);
+          (groupingTable.bindExternalIndices(&std::get<std::decay_t<Associated>>(associatedSlices)), ...);
+          auto binder = [&](auto&& x) {
+            (std::get<std::decay_t<Associated>>(associatedSlices).bindExternalIndices(&x), ...);
+          };
+          std::apply(
+            [&](auto&&... x) {
+              (binder(x), ...);
+            },
+            associatedSlices);
```

Comment:
Why not simply `apply(binder)`?

Reply:
This is a double loop - binder itself is folded over `Associated...` pack and is applied to a tuple. This allows to cross-bind all the indices in all requested tables.